loudml/prometheus.py

```diff
@@ -0,0 +1,392 @@
+        # values are pairs of [timestamp, value]
+        return self._response['data']['result'][0]['values']
+
+
+class PrometheusClient(object):
+    """
+    Client for Prometheus
+    """
+
+    def __init__(
+        self,
+        host="localhost",
+        port=9090,
+        ssl=False,
+        verify_ssl=False,
+        ssl_cert_path='',
+        user='',
+        password=''
+    ):
+        """
+        Set proper schema based on SSL param, open session
+        and supply basic authentication creds if given
+        https://requests.readthedocs.io/en/latest/user/advanced/#ssl-cert-verification
+        """
+        schema = "http"
+        if ssl:
+            schema = "https"
+
+        self.url = "%s://%s:%d" % (schema, host, port)
+        self.session = requests.Session()
+        if user and password:
+            self.session.auth = (user, password)
+        if ssl_cert_path:
+            self.session.verify = ssl_cert_path
+
+    def query(self, queries):
+        """
+        Run a list of queries against Prometheus.
+        API request should look like:
+        # http://0.0.0.0:9090/api/v1/query_range?query=go_memstats_alloc_bytes{instance=%22localhost:9090%22,%20job=%22prometheus%22}&start=1584783120&end=1584786720&step=15
+        # http://0.0.0.0:9090/api/v1/query_range?query=avg(go_memstats_alloc_bytes{instance=%22localhost:9090%22,%20job=%22prometheus%22})&start=1584783120&end=1584786720&step=15
+        """
+        if not isinstance(queries, list):
+            queries = [queries]
+
+        results = []
+
+        for q in queries:
+            time_pred = _build_time_predicates(q["start"], q["end"])
+            aggregator = AGGREGATORS.get(q["aggregator"])
+            if aggregator:
+                query_url = "{}/api/v1/query_range?{}&step={}&query={}({}{})".format(
+                    self.url,
+                    time_pred,
+                    q["step"],
+                    aggregator,
+                    q["metric_name"],
+                    q["tags"]
+                )
+            else:
+                query_url = "{}/api/v1/query_range?{}&step={}&query={}{}".format(
+                    self.url,
+                    time_pred,
+                    q["step"],
+                    q["metric_name"],
+                    q["tags"]
+                )
+
+            try:
+                resp = self.session.get(query_url)
+                self.session.close()
+            except requests.exceptions.SSLError as exn:
+                logging.error('Prometheus SSL error', str(exn))
+
+            if resp.ok:
+                results.append(PrometheusResult(resp.json()))
+            else:
+                logging.error('Prometheus error',
+                    resp.status_code, resp.text[:200])
+
+        return results
+
+
+    def put(self, entry):
+        """
+        Store a point in database
+        """
+        raise errors.NotImplemented("Prometheus bucket: not implemented yet")
```

Comment:
raise NotImplementedError

Reply:
agree. I also have seen a same in influx.py bucket